### PR TITLE
ISIS OTG model for Learned LSP.

### PR DESCRIPTION
**Pull request by SuryyaKrJana**
```json
{
  "choice": "ipv4_neighbors",
  "ipv4_neighbors": [
    {
      "ethernet_name": "string",
      "ipv4_address": "198.51.100.42",
      "link_layer_address": "string"
    }
  ],
  "ipv6_neighbors": [
    {
      "ethernet_name": "string",
      "ipv6_address": "2001:0db8:5b96:0000:0000:426f:8e17:642a",
      "link_layer_address": "string"
    }
  ],
  "bgp_prefixes": [
    {
      "bgp_peer_name": "string",
      "ipv4_unicast_prefixes": [
        {
          "ipv4_address": "string",
          "prefix_length": 0,
          "origin": "igp",
          "path_id": 0,
          "ipv4_next_hop": "198.51.100.42",
          "ipv6_next_hop": "2001:0db8:5b96:0000:0000:426f:8e17:642a",
          "communities": [
            {
              "type": "manual_as_number",
              "as_number": 65535,
              "as_custom": 65535
            }
          ],
          "as_path": {
            "segments": [
              {
                "type": "as_seq",
                "as_numbers": [
                  4294967295
                ]
              }
            ]
          }
        }
      ],
      "ipv6_unicast_prefixes": [
        {
          "ipv6_address": "string",
          "prefix_length": 0,
          "origin": "igp",
          "path_id": 0,
          "ipv4_next_hop": "198.51.100.42",
          "ipv6_next_hop": "2001:0db8:5b96:0000:0000:426f:8e17:642a",
          "communities": [
            {
              "type": "manual_as_number",
              "as_number": 65535,
              "as_custom": 65535
            }
          ],
          "as_path": {
            "segments": [
              {
                "type": "as_seq",
                "as_numbers": [
                  4294967295
                ]
              }
            ]
          }
        }
      ]
    }
  ],
  "isis_lsps": [
    {
      "isis_router_name": "string",
      "lsps": [
        {
          "lsp_id": "string",
          "pdu_type": "level_1",
          "remaining_lifetime": 4294967295,
          "sequence_number": 4294967295,
          "pdu_length": 65535,
          "flags": {
            "partition_repair": true,
            "attached_error": true,
            "attached_expense": true,
            "attached_delay": true,
            "attached_default": true,
            "overload": true
          },
          "is_type": 3,
          "tlvs": {
            "hostnames": [
              "string"
            ],
            "is_reachabilities": [
              {
                "neighbors": [
                  {
                    "system_id": "string"
                  }
                ]
              }
            ],
            "extended_is_reachabilities": [
              {
                "neighbors": [
                  {
                    "system_id": "string"
                  }
                ]
              }
            ],
            "ipv4_internal_reachabilities": [
              {
                "prefixes": [
                  {
                    "ipv4_address": "string",
                    "prefix_length": 0,
                    "redistribution_type": "up",
                    "default_metric": 63,
                    "origin_type": "internal"
                  }
                ]
              }
            ],
            "ipv4_external_reachabilities": [
              {
                "prefixes": [
                  {
                    "ipv4_address": "string",
                    "prefix_length": 0,
                    "redistribution_type": "up",
                    "default_metric": 63,
                    "origin_type": "internal"
                  }
                ]
              }
            ],
            "extended_ipv4_reachabilities": [
              {
                "prefixes": [
                  {
                    "ipv4_address": "198.51.100.42",
                    "prefix_length": 0,
                    "metric": 0,
                    "redistribution_type": "up",
                    "prefix_attributes": {
                      "x_flag": true,
                      "r_flag": true,
                      "n_flag": true
                    }
                  }
                ]
              }
            ],
            "ipv6_reachabilities": [
              {
                "prefixes": [
                  {
                    "ipv6_address": "2001:0db8:5b96:0000:0000:426f:8e17:642a",
                    "prefix_length": 0,
                    "metric": 0,
                    "redistribution_type": "up",
                    "origin_type": "internal",
                    "prefix_attributes": {
                      "x_flag": true,
                      "r_flag": true,
                      "n_flag": true
                    }
                  }
                ]
              }
            ]
          }
        }
      ]
    }
  ]
}

```

### CODE REVIEW

File: result/isislsp.yaml [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+            format: hex
+  
+    IsisLsps.State:
+      description: >-
+        ISIS LSP.
+      type: object
+      properties:
+        isis_router_name:
+          description: >-
+            The name of a ISIS Router.
```
Comment (ankur-sheth) [inline result/isislsp.yaml:33]: Grammar : name of a ISIS Router --> name of the ISIS Router.

Second, is this referring to the name that is configured on our side or the learnt name?

File: result/isislsp.yaml [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+        pdu_type:
+          description: >-
+            Link State PDU type.
+          type: string
+          enum: [level_1, level_2]
+        remaining_lifetime:
+          description: >-
+            Remaining lifetime in seconds before the LSP expiration.
+          type: integer
+          format: int64
```
Comment (ankur-sheth) [inline result/isislsp.yaml:59]: Why is this int64, yet the maximum value is 2^32?

File: result/isislsp.yaml [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+          minimum: 0
+          maximum: 4294967295
+        sequence_number:
+          description: >-
+            Sequence number of the LSP.
+          type: integer
+          format: int64
```
Comment (ankur-sheth) [inline result/isislsp.yaml:66]: Same comment as above.

File: result/isislsp.yaml [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+          minimum: 0
+          maximum: 65535
+        flags:
+          description: >-
+            LSP Type-Block flags.
+          $ref: '#/components/schemas/IsisLsp.Flags'
+        is_type:
+          description: >-
+            Type of neighboring system.
+          type: integer
```
Comment (ankur-sheth) [inline result/isislsp.yaml:82]: What is the significance of the values 1, 2 and 3?  Should this be an enum?

File: result/isislsp.yaml [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+          maximum: 3
+        hostname:
+          description: >-
+            This container defines TLV 137.
+          type: string
+        is_reachability:
+          description: |-
+            This container describes list of ISIS neighbors and attributes.
+          $ref: '#/components/schemas/IsisLsp.IsReachability'
+        extended_is_reachability:
```
Comment (ankur-sheth) [inline result/isislsp.yaml:93]: How is this field different than 'is_reachability'?  The descriptions of both are same, doesn't seem like they should be same.

File: result/isislsp.yaml [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+          description: |-
+            When set, the originator is attached to another area using the referred metric.
+          type: boolean
+        attached_expense:
+          description: |-
+            When set, the originator is attached to another
+            area using the referred metric.
+          type: boolean
+        attached_delay:
+          description: |-
```
Comment (ankur-sheth) [inline result/isislsp.yaml:136]: Same description as 'attached_expense'.  What does referred metric mean?

File: result/isislsp.yaml [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+          type: array
+          items:
+            $ref: '#/components/schemas/IsisLsp.ExtendedV4Prefix'
+
+    IsisLsp.ExtendedV4Prefix:
+      description: >-
+        This group defines attributes of an IPv4 standard prefix.
+      type: object
+      properties:
+        ipv4_address:
```
Comment (ankur-sheth) [inline result/isislsp.yaml:259]: Should this property have 'format : ipv4' in it?

File: result/isislsp.yaml [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+          type: array
+          items:
+            $ref: '#/components/schemas/IsisLsp.V6Prefix'
+
+    IsisLsp.V6Prefix:
+      description: >-
+        This group defines attributes of an IPv6 standard prefix.
+      type: object
+      properties:
+        ipv6_address:
```
Comment (ankur-sheth) [inline result/isislsp.yaml:296]: Should this property have 'format : ipv6' in it?

File: result/isislsp.yaml [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+          type: array
+          items:
+            $ref: '#/components/schemas/IsisLsp.V4Prefix'
+
+    IsisLsp.V4Prefix:
+      description: >-
+        This group defines attributes of an IPv4 standard prefix.
+      type: object
+      properties:
+        ipv4_address:
```
Comment (ankur-sheth) [inline result/isislsp.yaml:215]: Should this property have 'format : ipv4' in it?

**Review by ankur-sheth — CHANGES_REQUESTED** [review]
Several minor comments.  A few other comments that need clarifications.
Lastly, this will have to be updated once PR #255 is submitted.

File: result/isislsp.yaml (hunk 2 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:59]: thats the limitation of OTG, we raised earlier too. they support interger - 2^31-1 to + 2^31 + 1. to 2^32-1 max value, we have used int64 format as earlier too.

File: result/isislsp.yaml [hunk 10]
```diff
@@ -0,0 +1,341 @@
+components:
+  schemas:
+    IsisLsps.State.Request:
+      description: >-
+        The request to retrieve BGP peer prefix information.
```
Comment (SouravSinhaRoy) [inline result/isislsp.yaml:5]: ISIS lsp information

File: result/isislsp.yaml [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+      type: object
+      properties:
+        isis_router_names:
+          description: >-
+            The names of ISIS Routers to return results for.
+            An empty list will return results for all ISIS router.
```
[the remark this answers is not in the document]
Comment (SouravSinhaRoy) [reply result/isislsp.yaml:11]: The names of ISIS routers for which learned information is requested

File: result/isislsp.yaml [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+          type: string
+          format: hex
+        pdu_type:
+          description: >-
+            Link State PDU type.
+          type: string
+          enum: [level_1, level_2]
+        remaining_lifetime:
+          description: >-
+            Remaining lifetime in seconds before the LSP expiration.
```
Comment (SouravSinhaRoy) [inline result/isislsp.yaml:57]: before LSP expiration or before the LSP expires

File: result/isislsp.yaml [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+            When set, the originator is attached to another
+            area using the referred metric.
+          type: boolean
+        overload:
+          description: |-
+            When set, the originator is overloaded, and must
+            be avoided in path calculation.
+          type: boolean
+
+    IsisLsp.IsReachability:
```
Comment (SouravSinhaRoy) [inline result/isislsp.yaml:151]: IsReachability and externded is reachability describtions are same

Comment (SouravSinhaRoy) [reply result/isislsp.yaml:151]: descriptions i mean

File: result/isislsp.yaml [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+      description: >-
+        This container describes list of ISIS neighbors and attributes.
+      type: object
+      properties:
+        neighbors: 
+          description: >-
+            This container describes IS neighbors.
```
Comment (SouravSinhaRoy) [inline result/isislsp.yaml:158]: ISIS neighbors

File: result/isislsp.yaml (hunk 1 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:33]: This is name of ISIS the routers configured in our side that got learned LSP information.

File: result/isislsp.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (SuryyaKrJana) [reply result/isislsp.yaml:5]: done

File: result/isislsp.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (SuryyaKrJana) [reply result/isislsp.yaml:11]: done

File: result/isislsp.yaml [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+        isis_router_names:
+          description: >-
+            The names of ISIS Routers to return results for.
+            An empty list will return results for all ISIS router.
+          type: array
+          items:
+            type: string
+          x-constraint:
+            - "/components/schemas/Device.IsisRouter/properties/name"
+        lsp_filters:
```
[the remark this answers is not in the document]
Comment (SuryyaKrJana) [reply result/isislsp.yaml:17]: LSP ID is consist of the System ID of a neighbor, the Pseudonode ID, and the LSP number of the LSPs.
The last two bytes respectively represent Pseudonode ID and LSP number.
A pseudonode is a logical representation of the LAN which is generated by a Designated Intermediate System (DIS) on a LAN segment. If one LSP is not fit by maximum LSP size then thats sent in another LSP with LSP number is incremented by 1.
A router's learned LSP gets refreshed by 'remaining_lifetime'. Then the sequence number is incremented by 1.

File: result/isislsp.yaml (hunk 3 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:66]: same reply

File: result/isislsp.yaml (hunk 4 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:82]: updated with the comments.

File: result/isislsp.yaml (hunk 6 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:136]: Yes. Please referred to https://github.com/openconfig/public/blob/master/release/models/isis/openconfig-isis-lsp.yang.

File: result/isislsp.yaml (hunk 13 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:151]: updated with more details.

File: result/isislsp.yaml (hunk 9 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:215]: agrred.

File: result/isislsp.yaml (hunk 7 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:259]: agreed.

File: result/isislsp.yaml (hunk 8 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:296]: done.

File: result/isislsp.yaml (hunk 1 above)
Comment (ankur-sheth) [reply result/isislsp.yaml:39]: "name of a ISIS Router" --> "name of the ISIS Router"

**Review by ankur-sheth — CHANGES_REQUESTED** [review]
On grammar fix still pending, else ok.

File: result/isislsp.yaml [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+          type: array
+          items:
+            $ref: '#/components/schemas/IsisLsp.ExtendedV4Prefix'
+
+    IsisLsp.ExtendedV4Prefix:
+      description: >-
+        This group defines attributes of an IPv4 standard prefix.
+      type: object
+      properties:
+        ipv4_address:
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:268]: Do the prefix TLVs not need a mask ?

Comment (apratimmukherjee) [reply result/isislsp.yaml:268]: Also applies to gnmi spec,

File: result/isislsp.yaml [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+            $ref: '#/components/schemas/IsisLsp.V6Prefix'
+
+    IsisLsp.V6Prefix:
+      description: >-
+        This group defines attributes of an IPv6 standard prefix.
+      type: object
+      properties:
+        ipv6_address:
+          description: >-
+            An IPv6 address
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:308]: How is the mask made available for a prefix ( also applicable to gnmi )

File: result/isislsp.yaml (hunk 1 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:39]: done

File: result/isislsp.yaml [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+    IsisLsp.State:
+      description: >-
+        ISIS LSP.
+      type: object
+      properties:
+        lsp_id:
+          description: >-
+            LSP ID of the LSP, e.g. '6400000000010000'.
+          type: string
+          format: hex
```
[the remark this answers is not in the document]
Comment (SuryyaKrJana) [reply result/isislsp.yaml:55]: makes sense. shall do the sameformat in yang-model too?

File: result/isislsp.yaml [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+            Sequence number of the LSP.
+          type: integer
+          format: int64
+          minimum: 0
+          maximum: 4294967295
+        pdu_length:
+          description: >-
+            Total length of the LSP.
+          type: integer
+          minimum: 0
```
[the remark this answers is not in the document]
Comment (SuryyaKrJana) [reply result/isislsp.yaml:79]: true, but as a model we can let to know the range of a field by mentioning the min & max value.

File: result/isislsp.yaml [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+            IS Type - bits 1 and 2 indicate the type of Intermediate System.
+            1 - ( i.e. bit 1 set) Level 1 Intermediate system.
+            2 - Unused value.
+            3 - (i.e. bits 1 and 2 set) Level 2 Intermediate system. 
+          type: integer
+          minimum: 1
+          maximum: 3
+        hostname:
+          description: >-
+            This container defines TLV 137.
```
Comment (PrasenjitAdhikary) [inline result/isislsp.yaml:96]: Not a container. Also, the yang model shows a list of host name(s) - so either of the models should be corrected for consistency.

Comment (SuryyaKrJana) [reply result/isislsp.yaml:96]: ISIS Yang-model referes as array. But our isis router can have  only hostname. will change to non-array there.

File: result/isislsp.yaml (hunk 16 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:268]: missed somehow.

File: result/isislsp.yaml (hunk 17 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:308]: will correct there too.

File: result/isislsp.yaml (hunk 14 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:167]: Its ok. IS neighbor (Intermediate System neighbor)

File: result/isislsp.yaml [hunk 21]
```diff
@@ -0,0 +1,341 @@
...
+            The names of ISIS Routers to return results for.
+            An empty list will return results for all ISIS router.
+          type: array
+          items:
+            type: string
+          x-constraint:
+            - "/components/schemas/Device.IsisRouter/properties/name"
+        lsp_filters:
+          description: >-
+            Specify which LSPs to return. If the list is empty or missing
```
[the remark this answers is not in the document]
Comment (SuryyaKrJana) [reply result/isislsp.yaml]: we have removed the filter.

File: result/isislsp.yaml (hunk 20 above)
Comment (apratimmukherjee) [reply result/isislsp.yaml:89]: Can we check the RFC/spec . If it allows multiple we should keep as multiple. Maybe send only first name initially . When fixed and if spec allows, send multiple. 
But should be spec driven and not our implementation,

File: result/isislsp.yaml (hunk 14 above)
Comment (apratimmukherjee) [reply result/isislsp.yaml:160]: Maybe change it then to Intermediate System neighbors explicitly. Others might have similar confusion when reading this,

File: result/isislsp.yaml [hunk 22]
```diff
@@ -36,9 +22,11 @@ components:
       properties:
         isis_router_name:
           description: >-
-            The name of a ISIS Router.
+            The name of the ISIS Router.
           type: string
         lsps:
+          description: >-
+            One more more LSPs learned by this ISIS router.
```
Comment (ankur-sheth) [inline result/isislsp.yaml:29]: Something is wrong in this description.

File: result/isislsp.yaml [hunk 23]
```diff
@@ -50,9 +38,14 @@ components:
       properties:
         lsp_id:
           description: >-
-            LSP ID of the LSP, e.g. '6400000000010000'.
+            LSP ID of the LSP, in the format, e.g. '640000000001-00-00'.
+            LSP ID is consist of the System ID of a neighbor, the Pseudonode ID, and the LSP number of the LSPs.
```
Comment (ankur-sheth) [inline result/isislsp.yaml:42]: LSP ID is consist of --> LSP ID consists of
and the LSP number of the LSPs --> and the LSP number.

File: result/isislsp.yaml [hunk 24]
```diff
@@ -50,9 +38,14 @@ components:
...
+            The last two bytes respectively represent Pseudonode ID and LSP number.
```
Comment (ankur-sheth) [inline result/isislsp.yaml:43]: The lsat two bytes represent Pseudonode ID and LSP number respectively.

File: result/isislsp.yaml [hunk 25]
```diff
@@ -50,9 +38,14 @@ components:
...
+            A pseudonode is a logical representation of the LAN which is generated by a Designated Intermediate System (DIS)
+            on a LAN segment. If one LSP is not fit by maximum LSP size then thats sent in another LSP with LSP number
```
Comment (ankur-sheth) [inline result/isislsp.yaml:45]: If one LSP exceeds the maximum LSP size then it is sent in another LSP with the LSP number incremented by one.

File: result/isislsp.yaml [hunk 26]
```diff
@@ -267,9 +260,13 @@ components:
       properties:
         ipv4_address:
           description: >-
-            An IPv4 unicast address
+            An IPv4 unicast address.
           type: string
           format: ipv4
+        prefix_length:
+          description:
+            The length of the prefix.
```
Comment (ankur-sheth) [inline result/isislsp.yaml:268]: The length of the IPv4 prefix.

File: result/isislsp.yaml [hunk 27]
```diff
@@ -308,6 +305,10 @@ components:
             An IPv6 address
           type: string
           format: ipv6
+        prefix_length:
+          description:
+            The length of the prefix.
```
Comment (ankur-sheth) [inline result/isislsp.yaml:310]: The length of the IPv6 prefix.

**Review by ankur-sheth — CHANGES_REQUESTED** [review]

**Review by ankur-sheth — APPROVED** [review]

File: result/states.yaml [hunk 28]
```diff
@@ -11,13 +11,16 @@ components:
...
           - bgp_prefixes
+          - isis_lsp_states
           default: ipv4_neighbors
         ipv4_neighbors:
           $ref: './neighborsv4.yaml#/components/schemas/Neighborsv4.States.Request'
         ipv6_neighbors:
           $ref: './neighborsv6.yaml#/components/schemas/Neighborsv6.States.Request'
         bgp_prefixes:
           $ref: './bgpprefix.yaml#/components/schemas/BgpPrefix.State.Request'
+        isis_lsp_states:
```
Comment (ashutshkumr) [inline result/states.yaml:22]: Isn't `isis_lsps` a more appropriate name here ?

File: result/isislsp.yaml [hunk 29]
```diff
@@ -0,0 +1,356 @@
+components:
+  schemas:
+    IsisLsps.State.Request:
+      description: >- 
+        The request to retrieve ISIS Link State Protocol Data Unit (LSP) information learned by this router.
```
Comment (ashutshkumr) [inline result/isislsp.yaml:5]: - Should be "ISIS Link State PDUs (LSP)" - for same reason why we didn't expand ISIS
- Correction - "information learned by the routers"

File: result/isislsp.yaml [hunk 30]
```diff
@@ -0,0 +1,356 @@
...
+            $ref: '#/components/schemas/IsisLsp.State'
+
+    IsisLsp.State:
+      description: >-
+        ISIS LSP.
+      type: object
+      properties:
+        lsp_id:
+          description: >-
+            LSP ID of the LSP, in the format, e.g. '640000000001-00-00'.
```
Comment (ashutshkumr) [inline result/isislsp.yaml:41]: "LSP ID of the LSP, in the format" -> "LSP ID in the format"

Comment (SuryyaKrJana) [reply result/isislsp.yaml:41]: ok.

**Review by ashutshkumr — APPROVED** [review]

File: result/states.yaml (hunk 28 above)
Comment (SuryyaKrJana) [reply result/states.yaml:22]: ok.

File: result/isislsp.yaml [hunk 31]
```diff
@@ -0,0 +1,380 @@
...
+          minimum: 1
+          maximum: 3
+        tlvs:
+          description: |-
+            This container defines Link State PDU State TLVs.
+          $ref: '#/components/schemas/IsisLsp.Tlvs'
+
+    IsisLsp.Tlvs:
+      description: >-
+        This contains all TLVs. As per ISIS-Yang model Hostname, IS-Reachability, Extended-IS-Reachability etc. are list.
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:93]: Please remove second line . OTG descriptions should be standalone else much more detail is needed as well as references etc. to the yang model to understand 2nd line.

File: result/isislsp.yaml [hunk 32]
```diff
@@ -0,0 +1,380 @@
...
+      type: object
+      properties:
+        hostnames:
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:96]: Did we not agree to have a list of hostname TLVs and hostnames under that ?

File: result/isislsp.yaml [hunk 33]
```diff
@@ -0,0 +1,380 @@
...
+          description: >-
+            This container defines TLV 137 for the hostname.
+          type: array
+          items:
+            type: string
+        is_reachabilities:
+          description: |-
+            Array of list of ISIS neighbors and attributes.
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:104]: All of these : Did we not agree to have a list of TLVs of each time with each again having an array of information when needed ?

File: result/isislsp.yaml [hunk 34]
```diff
@@ -0,0 +1,380 @@
...
+            $ref: '#/components/schemas/IsisLsp.IsReachability'
+        extended_is_reachabilities:
+          description: |-
+            Array of list of ISIS neighbors and attributes.
+          type: array
+          items:
+            $ref: '#/components/schemas/IsisLsp.ExtendedIsReachability'
+        ipv4_internal_reachabilities:
+          description: |-
+            Array of list of ISIS neighbors and attributes. list of IPv4 internal reachability information.
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:116]: I thought at this level we should be having array of TLVs of type ipv4_internal_reachability ?

**Review by apratimmukherjee — COMMENTED** [review]
I am not seeing the list of TLVs as we discussed for OTG,

File: result/isislsp.yaml [hunk 35]
```diff
@@ -0,0 +1,380 @@
...
+
+    IsisLsp.Ipv4InternalReachability:
+      description: |-
+        This container defines list of IPv4 internal reachability information.
+        This is advertised when the origin-type is set 'internal' in route range configurations.
+      type: object
+      properties:
+        prefixes:
+          description: >-
+            This container describes IS neighbors.
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:216]: describes list of  IPv4 prefixes in this TLV.

File: result/isislsp.yaml [hunk 36]
```diff
@@ -0,0 +1,380 @@
...
+      properties:
+        system_id:
+          description: >-
+            The System ID for this emulated ISIS router, e.g. "640100010000".
+          type: string
+          format: hex
+
+    IsisLsp.Ipv4InternalReachability:
+      description: |-
+        This container defines list of IPv4 internal reachability information.
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:210]: This container defines the information present on one IPv4 Internal Reachability TLV.

File: result/isislsp.yaml (hunk 34 above)
Comment (apratimmukherjee) [reply result/isislsp.yaml:116]: Array of list of IPv4 Internal Reachability TLVs present in this LSP.

File: result/isislsp.yaml [hunk 37]
```diff
@@ -0,0 +1,390 @@
...
+    IsisLsp.V4Prefix:
+      description: >-
+        This group defines attributes of an IPv4 standard prefix.
+      type: object
+      properties:
+        ipv4_address:
+          description: >-
+            An IPv4 address.
+          type: string
+        prefix_length:
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:253]: If we agree on min-max as being necessary for LI fields, then prefix lengths for v4/v6 should have min max of 0/32 and 0/128 respectively,

File: result/isislsp.yaml [hunk 38]
```diff
@@ -0,0 +1,390 @@
...
+          type: string
+          format: ipv4
+        prefix_length:
+          description:
+            The length of the IPv4 prefix.
+          type: integer
+        metric:
+          description: >-
+            ISIS wide metric.
+          type: integer
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:310]: min-max ?

File: result/isislsp.yaml [hunk 39]
```diff
@@ -0,0 +1,390 @@
...
+      properties:
+        ipv6_address:
+          description: >-
+            An IPv6 address
+          type: string
+          format: ipv6
+        prefix_length:
+          description:
+            The length of the IPv6 prefix.
+          type: integer
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:348]: min/max ?

File: result/isislsp.yaml [hunk 40]
```diff
@@ -0,0 +1,390 @@
...
+        metric:
+          description: >-
+            ISIS wide metric.
+          type: integer
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:352]: min/max ?

File: result/isislsp.yaml [hunk 41]
```diff
@@ -0,0 +1,390 @@
...
+            $ref: '#/components/schemas/IsisLsp.ExtendedV4Prefix'
+
+    IsisLsp.ExtendedV4Prefix:
+      description: >-
+        This group defines attributes of an IPv4 standard prefix.
+      type: object
+      properties:
+        ipv4_address:
+          description: >-
+            An IPv4 unicast address.
```
[the remark this answers is not in the document]
Comment (SuryyaKrJana) [reply result/isislsp.yaml:300]: This comments copied from bgp section :)

File: result/isislsp.yaml [hunk 42]
```diff
@@ -0,0 +1,380 @@
...
+            Array of Hostname TLVs ( type 137) present in this LSP.
+          type: array
+          items:
+            $ref: '#/components/schemas/IsisLsp.Hostname'
+        is_reachability_tlvs:
+          description: |-
+            Array of IS-Reachability TLVs (type 2) present in this LSP.
+          type: array
+          items:
+            $ref: '#/components/schemas/IsisLsp.IsReachability'
```
Comment (PrasenjitAdhikary) [inline result/isislsp.yaml:99]: Can we also rename all child object names, for consistency with the attribute name? IsReachability --> IsReachabilityTlv

File: result/isislsp.yaml [hunk 43]
```diff
@@ -0,0 +1,380 @@
...
+            on a LAN segment. If one LSP exceeds the maximum LSP size then it is sent in another LSP with the LSP number incremented by one.
+            A router's learned LSP gets refreshed by 'remaining_lifetime'. Then the sequence number is incremented by 1.
+          type: string
+        pdu_type:
+          description: >-
+            Link State PDU type.
+          type: string
+          enum: [level_1, level_2]
+        remaining_lifetime:
+          description: >-
```
Comment (PrasenjitAdhikary) [inline result/isislsp.yaml:54]: properties marked "mandatory" in yang should be indicated with "required" attribute to indicate MUST in the otg-models as well.

**Review by PrasenjitAdhikary — APPROVED** [review]
Looks good to me. Need to address some object renaming and "required" attributes before check-in.

File: result/isislsp.yaml (hunk 40 above)
Comment (SuryyaKrJana) [reply result/isislsp.yaml:342]: removed min/max from other places

File: result/isislsp.yaml [hunk 44]
```diff
@@ -0,0 +1,381 @@
...
+            The origin of the advertised route-internal or external to the ISIS area.
+            Options include the following: 
+              Internal-for intra-area routes, through Level 1 LSPs. 
+              External-for inter-area routes redistributed within L1, through Level 1 LSPs.
+          type: string
+          enum: [internal, external]
+  
+    IsisLsp.ExtendedIpv4ReachabilityTlv:
+      description: |-
+        This container defines list of IPv4 extended reachability information.
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:273]: in one Extended IPv4 External Reachability TLV.

File: result/isislsp.yaml [hunk 45]
```diff
@@ -0,0 +1,381 @@
...
+            The prefixes are being advertised from a higher level (L2) down to a lower level (L1). 
+          type: string
+          enum: [up, down]
+        prefix_attributes:
+          $ref: '#/components/schemas/IsisLsp.PrefixAttributes'
+        
+
+    IsisLsp.Ipv6ReachabilityTlv:
+      description: |-
+        It defines list of IPv6 extended reachability information.
```
Comment (apratimmukherjee) [inline result/isislsp.yaml:316]: in one IPv6 Reachability TLV.

**Review by apratimmukherjee — APPROVED** [review]
Two minor improvements mentioned for documentation. Rest  LGTM.

**Comment by SuryyaKrJana** [comment]
Merging ISIS  OTG Telemetry to master

**Pull request merged**